cpp/src/sampling/uniform_neighbor_sampling_impl.hpp

```diff
@@ -115,13 +115,165 @@ uniform_nbr_sample_impl(
+           rmm::device_uvector<int32_t>,
+           std::optional<rmm::device_uvector<int32_t>>>
+uniform_neighbor_sample_impl(
+  raft::handle_t const& handle,
+  graph_view_t<vertex_t, edge_t, store_transposed, multi_gpu> const& graph_view,
+  std::optional<edge_property_view_t<edge_t, weight_t const*>> edge_weight_view,
+  std::optional<
+    edge_property_view_t<edge_t,
+                         thrust::zip_iterator<thrust::tuple<edge_t const*, edge_type_t const*>>>>
+    edge_type_view,
+  rmm::device_uvector<vertex_t>& d_in,
+  std::optional<rmm::device_uvector<int32_t>>& d_labels,
+  raft::host_span<const int> h_fan_out,
+  bool with_replacement,
+  uint64_t seed)
+{
+#ifdef NO_CUGRAPH_OPS
+  CUGRAPH_FAIL(
+    "uniform_neighbor_sampl_impl not supported in this configuration, built with NO_CUGRAPH_OPS");
+#else
+
+  CUGRAPH_EXPECTS(h_fan_out.size() > 0,
+                  "Invalid input argument: number of levels must be non-zero.");
+
+  rmm::device_uvector<vertex_t> d_result_src(0, handle.get_stream());
+  rmm::device_uvector<vertex_t> d_result_dst(0, handle.get_stream());
+  rmm::device_uvector<edge_t> d_result_edge_id(0, handle.get_stream());
+  auto d_result_weight =
+    edge_weight_view ? std::make_optional(rmm::device_uvector<weight_t>(0, handle.get_stream()))
+                     : std::nullopt;
+  auto d_result_edge_type =
+    edge_type_view ? std::make_optional(rmm::device_uvector<edge_type_t>(0, handle.get_stream()))
+                   : std::nullopt;
+  rmm::device_uvector<int32_t> d_result_hop(0, handle.get_stream());
+  auto d_result_label = d_labels
+                          ? std::make_optional(rmm::device_uvector<int32_t>(0, handle.get_stream()))
+                          : std::nullopt;
+
+  int32_t hop{0};
+  size_t comm_size{1};
+
+  if constexpr (multi_gpu) {
+    seed += handle.get_comms().get_rank();
+    comm_size = handle.get_comms().get_size();
+  }
+
+  for (auto&& k_level : h_fan_out) {
+    // prep step for extracting out-degs(sources):
+    if constexpr (multi_gpu) {
+      if (d_labels) {
+        std::tie(d_in, *d_labels) = shuffle_int_vertices_and_values_by_gpu_id(
+          handle, std::move(d_in), std::move(*d_labels), graph_view.vertex_partition_range_lasts());
+      } else {
+        d_in = shuffle_int_vertices_by_gpu_id(
+          handle, std::move(d_in), graph_view.vertex_partition_range_lasts());
+      }
+    }
+
+    rmm::device_uvector<vertex_t> d_out_src(0, handle.get_stream());
+    rmm::device_uvector<vertex_t> d_out_dst(0, handle.get_stream());
+    rmm::device_uvector<edge_t> d_out_edge_id(0, handle.get_stream());
+    std::optional<rmm::device_uvector<weight_t>> d_out_weight{std::nullopt};
+    std::optional<rmm::device_uvector<edge_type_t>> d_out_edge_type{std::nullopt};
+    std::optional<rmm::device_uvector<int32_t>> d_out_label{std::nullopt};
+
+    if (k_level > 0) {
+      raft::random::RngState rng_state(seed);
+      seed += d_in.size() * k_level * comm_size;
```

Comment:
Why not just define `rng_state` outside the outermost loop?

Reply:
We could.  I'll look at that for my next update.